loader/src/references/reference_binder.py

```diff
@@ -0,0 +1,137 @@
+        reference_attribute = fhir_object[reference_path]
+
+        # If we have a list of references, we want to bind all of them.
+        # Thus, we loop on all the items in sub_fhir_object.
+        if not isinstance(reference_attribute, list):
+            reference_attribute = [reference_attribute]
+
+        for ref in reference_attribute:
+            # extract the type and itentifier of the reference
+            reference_type = ref["type"]
+            identifier = ref["identifier"]
+            if not identifier.get("value") or not identifier.get("system"):
+                logger.error(
+                    f"invalid reference: {ref}. identifier.value and identifier.system are required"
+                )
+                continue
+
+            # search the referenced resource in the database
+            referenced_resource = self.fhirstore.db[reference_type].find_one(
+                {"identifier": {"value": identifier["value"], " system": identifier["system"]}},
+                ["id"],
+            )
+            if referenced_resource:
+                # if found, add the ID as the "literal reference"
+                # (https://www.hl7.org/fhir/references-definitions.html#Reference.reference)
+                logger.info(f"reference to {reference_type} {identifier['value']} resolved")
+                ref["reference"] = f"{reference_type}/{referenced_resource['id']}"
+            else:
+                logger.info(
+                    f"caching reference to {reference_type} "
+                    f"{identifier['value']} at {reference_path}"
+                )
+                # otherwise, cache the reference to resolve it later
+                cache_key = (reference_type, identifier["system"], identifier["value"])
+                pending_refs = self.cache.get(cache_key, {})
+                self.cache[cache_key] = {
+                    **pending_refs,
+                    (fhir_object["resourceType"], reference_path): [
+                        *pending_refs.get(fhir_object["resourceType"], []),
+                        fhir_object["id"],
+                    ],
+                }
+
+        return reference_attribute
+
+    def resolve_pending_references(self, fhir_object):
+        for identifier in fhir_object["identifier"]:
+            if not identifier.get("value") or not identifier.get("system"):
+                logger.error(
+                    f"invalid identifier: {identifier}. "
+                    "identifier.value and identifier.system are required"
+                )
+                continue
+            cache_key = (fhir_object["resourceType"], identifier["system"], identifier["value"])
+            pending_refs = self.cache.get(cache_key, {})
+            for (source_type, reference_path), refs in pending_refs.items():
+                find_predicate = {
+                    "id": {"$in": refs},
+                    f"{reference_path}.identifier.value": identifier["value"],
```

Comment:
Why don't you use the system in the search?

Reply:
my bad. done.